### ISSUE TITLE
Test if ZongJi supports binlog_row_image != full

### ISSUE THREAD

**Issue by numtel** [nevill/zongji#51]
http://dev.mysql.com/doc/refman/5.6/en/replication-options-binary-log.html#sysvar_binlog_row_image

**Comment by vlasky** [comment]
Thanks for posting this! I was unaware of the introduction of the binlog_row_image variable in MySQL 5.6.2.

Some time back we actually changed our schemas to put BLOBs in a separate table so that large amounts of unchanged data would not be wastefully written in the binary log, but with the 'minimal' or 'noblob' options that could have been avoided.

Would be very interested to know if ZongJi does indeed support this, as I expect this would also reduce CPU usage required to read the binary log and improve application scalability.

**Comment by vlasky** [comment]
I confirm that ZongJi currently does NOT support binlog_row_image != full.

I performed a test in which I set binlog_row_image=noblob so that BLOB and TEXT column values would be omitted from rows written to the MySQL binlog unless those values actually changed in the affected rows.

A **RangeError: Index out of range** exception was generated within the [mysql module](https://github.com/mysqljs/mysql/) binlog parser code:

```
Oct 15 11:31:35 myserver node[3256]: /var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/node_modules/mysql/lib/protocol/Parser.js:80
Oct 15 11:31:35 myserver node[3256]: throw err; // Rethrow non-MySQL errors
Oct 15 11:31:35 myserver node[3256]: ^
Oct 15 11:31:35 myserver node[3256]: RangeError: Index out of range
Oct 15 11:31:35 myserver node[3256]: at checkOffset (buffer.js:977:11)
Oct 15 11:31:35 myserver node[3256]: at Uint8Array.Buffer.readInt32BE (buffer.js:1142:5)
Oct 15 11:31:35 myserver node[3256]: at readIntBE (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/lib/common.js:198:24)
Oct 15 11:31:35 myserver node[3256]: at Object.exports.readMysqlValue (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/lib/common.js:559:13)
Oct 15 11:31:35 myserver node[3256]: at readRow (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/lib/rows_event.js:111:16)
Oct 15 11:31:35 myserver node[3256]: at UpdateRows._fetchOneRow (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/lib/rows_event.js:143:13)
Oct 15 11:31:35 myserver node[3256]: at UpdateRows.RowsEvent (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/lib/rows_event.js:62:27)
Oct 15 11:31:35 myserver node[3256]: at new UpdateRows (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/lib/rows_event.js:135:13)
Oct 15 11:31:35 myserver node[3256]: at BinlogHeader.parse (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/lib/packet/binlog_header.js:47:23)
Oct 15 11:31:35 myserver node[3256]: at Protocol._parsePacket (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/node_modules/mysql/lib/protocol/Protocol.js:261:12)
Oct 15 11:31:35 myserver node[3256]: at Parser.write (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/node_modules/mysql/lib/protocol/Parser.js:76:12)
Oct 15 11:31:35 myserver node[3256]: at Protocol.write (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/node_modules/mysql/lib/protocol/Protocol.js:39:16)
Oct 15 11:31:35 myserver node[3256]: at Socket.<anonymous> (/var/www/html/meteor/prod-20191015-104100/bundle/programs/server/npm/node_modules/meteor/vlasky_mysql/node_modules/zongji/node_modules/mysql/lib/Connection.js:103:28)
Oct 15 11:31:35 myserver node[3256]: at emitOne (events.js:116:13)
Oct 15 11:31:35 myserver node[3256]: at Socket.emit (events.js:211:7)
Oct 15 11:31:35 myserver node[3256]: at Socket.Readable.read (_stream_readable.js:475:10)
Oct 15 11:31:35 myserver systemd[1]: dispatch.service: main process exited, code=exited, status=1/FAILURE
Oct 15 11:31:35 myserver systemd[1]: Unit dispatch.service entered failed state.  
```

**Comment by vlasky** [comment]
I think the code that needs to be tweaked is in [readRow() in lib/rows_event line 96](https://github.com/nevill/zongji/blob/42f43d54bd6ec3f8a47d8eb52cc1cbcbc3077b0a/lib/rows_event.js#L96).

Also, I found this excellent document [Dive into MySQL replication protocol](https://medium.com/@siddontang/dive-into-mysql-replication-protocol-cd14791bcc  ) which has some pertinent information on the handling of `binlog_row_image=minimal` and `binlog_row_image=noblob`:
 

> From MySQL 5.6, it supports another two binlog row images: minimal and noblob. For minimal row image update, if we have 16 columns and only the first column data changed, if we use (num of columns+7)/8, we should use 2 bytes to store null bitmap, but if we use (bits set in ‘columns-present-bitmap’+7)/8, we will only use 1 bytes to store null bitmap, saving 1 byte(is it really necessary?). By the way, I sent a [pull request](https://github.com/noplay/python-mysql-replication/pull/103) for [python-mysql-replication](https://github.com/noplay/python-mysql-replication) to handle minimal and noblob row image paring.
> 
> Now we get column-present-bitmap and null-bitmap, for a column, if it’s not set in column-present-bitmap or set in null-bitmap, we will know that this column is null for the current row data.
> 
> Then we will parse the rest of none null columns. For some special columns, like MYSQL_TYPE_LONG or MYSQL_TYPE_FLOAT, we can know the data length directly, e.g, MYSQL_TYPE_LONG is 4 bytes and MYSQL_TYPE_TINY_INT is 1 byte.
> 
> But for other columns, we should use column meta in TABLE_MAP_EVENT to help us determine the data length. For example, a MYSQL_TYPE_BLOB column, if meta is 1, the data is tiny blob and the first 1 byte in data is the length for payload, if meta is 2, the data is short blob and the first 2 bytes is the length for payload.
> 
> Paring the real data is very hard and I can not illustrate here fully and clearly. I hope you can see the source in MySQL or [go-mysql](https://github.com/siddontang/go-mysql) replication pkg if you have some interest.

**Comment by nevill** [comment]
This is fixed in v0.5.1.
